contracts/SimpleGAMultiSig.aes

```diff
@@ -0,0 +1,153 @@
+// ISC License
+//
+// Copyright (c) 2022, aeternity developers
+//
+// Credit to Michel Meier & Marco Walz (kryptokrauts.com)
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
+// REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+// AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
+// INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+// LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE
+// OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+// PERFORMANCE OF THIS SOFTWARE.
+
+// THIS IS NOT SECURITY AUDITED
+// DO NEVER USE THIS WITHOUT SECURITY AUDIT FIRST
+
+@compiler >= 6
+
+include "Option.aes"
+include "String.aes"
+
+// This contract provides multi signature support based on generalized accounts
+// A transaction can be proposed by one of the multisig users which are registered during contract deployment
+// The transaction must be confirmed by state.confirmations_required of the registered multisig users, while the proposing user automatically confirms
+// If the amount of confirmations is reached, authenticate can be called by any of the multisig users and the original tx is processed
+// Currently only one proposed transaction can be multisigned at a time
+contract SimpleGAMultiSig =
+
+  datatype event
+    = TxProposed(hash, address, int)
+    | TxConfirmed(hash, address)
+    | TxConsensusReached(hash)
+    | TxAuthorized(hash)
+    | TxRevoked(address)
+
+  record transaction = 
+    { ga_tx_hash : hash,
+      expires_at : int,
+      proposed_by : address,
+      confirmed_by : list(address) }
+
+  record state = 
+  	{ multisig_users : list(address),
+  	  confirmations_required : int,    
+      current_tx : option(transaction),
+      nonce : int,
+      owner: address }
+
+  stateful entrypoint init(p_confirmations_required : int, p_multisig_users : list(address)) = 
+    { multisig_users = Call.caller :: p_multisig_users, 
+  	  confirmations_required = p_confirmations_required,
+  	  current_tx = None,
+  	  nonce = 1,
+      owner = Call.caller }
+
+  // GA authorize function
+  stateful entrypoint authorize(sig: signature, n : int) : bool =  
+    check_nonce(n)
+    switch(is_expired())
+      true => 
+        abort("ERROR_TX_EXPIRED")
+      false =>
+        require(has_consensus(), "ERROR_NO_CONSENSUS")
```

Comment:
if this returns the `current_tx` then in the next steps you don't have to `Option.force`

Reply:
can you give a short example how this should work in this specific scenario? the `has_consensus` function is also used at another placed to emit an event after consensus has been reached.

do I understand correctly that you propose to move the `require` statement into the function and instead of returning a boolean return `current_tx` ?

Reply:
what about this? (still uses `Option.force` in advance and directly performs the consensus check)

```
let tx = Option.force_msg(state.current_tx, "ERROR_NO_TX_PROPOSED")
require(List.length(tx.confirmed_by) >= state.confirmations_required, "ERROR_NO_CONSENSUS")
```

Reply:
not sure how we get advantage with your proposal and how it could be implemented